{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bee097d7",
   "metadata": {},
   "source": [
    "## This notebook accompanies our submission for SWAT4HCLS 2023 \"data-in-use\", showcasing federated queries across UniProt, OMA and Bgee for identifying essential proteins in Yeast\n",
    "\n",
    "Towards reproducing the analysis in the paper:\n",
    "Zhang, Xue, Wangxin Xiao, and Xihao Hu. \"Predicting essential proteins by integrating orthology, gene expressions, and PPI networks.\" PloS one 13.4 (2018): e0195410.\n",
    "\n",
    "#### Note: we build the query progressively to illustrate step by step how to achieve the end result. We also highlight lessons learned for each of the steps, where relevant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d1bc30df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: SPARQLWrapper in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (1.8.5)\r\n",
      "Requirement already satisfied: rdflib>=4.0 in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from SPARQLWrapper) (6.2.0)\r\n",
      "Requirement already satisfied: isodate in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from rdflib>=4.0->SPARQLWrapper) (0.6.1)\r\n",
      "Requirement already satisfied: setuptools in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from rdflib>=4.0->SPARQLWrapper) (58.0.4)\r\n",
      "Requirement already satisfied: pyparsing in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from rdflib>=4.0->SPARQLWrapper) (3.0.4)\r\n",
      "Requirement already satisfied: six in /Users/asima/opt/anaconda3/lib/python3.9/site-packages (from isodate->rdflib>=4.0->SPARQLWrapper) (1.16.0)\r\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5d/3j6fyxq90gs2qdxgzpzsl1gh0000gp/T/ipykernel_54782/2522515982.py:9: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.\n",
      "  pd.set_option('display.max_colwidth', -1)\n"
     ]
    }
   ],
   "source": [
    "# helper functions\n",
    "import sys\n",
    "!{sys.executable} -m pip install SPARQLWrapper\n",
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "import sys, os, time\n",
    "import pandas as pd\n",
    "\n",
    "# always display full column results (don't truncate output)\n",
    "pd.set_option('display.max_colwidth', -1)\n",
    "\n",
    "OMA_SPARQL_endpoint = \"https://sparql.omabrowser.org/sparql\"\n",
    "Bgee_SPARQL_endpoint = \"https://bgee.org/sparql/\"\n",
    "UniProt_SPARQL_endpoint = \"https://sparql.uniprot.org/sparql/\"\n",
    "\n",
    "# the endpoints must be defined as wrappers for executing SPARQL queries\n",
    "sparql_endpoint_OMA = SPARQLWrapper(OMA_SPARQL_endpoint)\n",
    "sparql_endpoint_UniProt = SPARQLWrapper(UniProt_SPARQL_endpoint)\n",
    "\n",
    "# function to print in a table results of a SPARQL query\n",
    "def pretty_print(results):\n",
    "    \n",
    "    # how to transform SPARQL results into Pandas dataframes\n",
    "    \n",
    "    # get header (column names) from results\n",
    "    header = results[\"results\"][\"bindings\"][0].keys()\n",
    "\n",
    "    # display table of results:\n",
    "    table = []\n",
    "    \n",
    "    # the SPARQL JSON results to the query are available in the \"results\", \"bindings\" entry:\n",
    "    for entry in results[\"results\"][\"bindings\"]:\n",
    "        # append entries from the results to a regular Python list of rows, which we can then transform to a Pandas DF\n",
    "        row = [entry[column][\"value\"] if entry.get(column, None) != None else None for column in header]\n",
    "        table.append(row)\n",
    "    df = pd.DataFrame(table, columns=list(header))\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdfa82a2",
   "metadata": {},
   "source": [
    "### 1. Identify the taxon ID of Yeast (\"S. cerevisiae\") in OMA RDF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca11041e",
   "metadata": {},
   "source": [
    "#### First attempt: identify this directly in OMA, searching by scientific name of Yeast (Saccharomyces Cerevisiae)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9ca68031",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_OMA_s_cerevisiae = \"\"\"\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "\n",
    "select ?taxon ?name where {\n",
    "    ?taxon  a up:Taxon.\n",
    "    ?taxon up:scientificName ?name.\n",
    "    filter (strstarts (lcase(?name), \"saccharomyces cerevisiae\" ))\n",
    "}\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a50c7824",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/559292</td>\n",
       "      <td>Saccharomyces cerevisiae (strain ATCC 204508 / S288c)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764097</td>\n",
       "      <td>Saccharomyces cerevisiae (strain AWRI796)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764099</td>\n",
       "      <td>Saccharomyces cerevisiae (strain VIN 13)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764101</td>\n",
       "      <td>Saccharomyces cerevisiae (strain FostersO)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     taxon  \\\n",
       "0  http://purl.uniprot.org/taxonomy/559292   \n",
       "1  http://purl.uniprot.org/taxonomy/764097   \n",
       "2  http://purl.uniprot.org/taxonomy/764099   \n",
       "3  http://purl.uniprot.org/taxonomy/764101   \n",
       "\n",
       "                                                    name  \n",
       "0  Saccharomyces cerevisiae (strain ATCC 204508 / S288c)  \n",
       "1  Saccharomyces cerevisiae (strain AWRI796)              \n",
       "2  Saccharomyces cerevisiae (strain VIN 13)               \n",
       "3  Saccharomyces cerevisiae (strain FostersO)             "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# set the query to be executed against the OMA endpoint and set the return format to JSON\n",
    "sparql_endpoint_OMA.setQuery(query_OMA_s_cerevisiae)\n",
    "sparql_endpoint_OMA.setReturnFormat(JSON)\n",
    "\n",
    "results = sparql_endpoint_OMA.query().convert()\n",
    "\n",
    "pretty_print(results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eea08b51",
   "metadata": {},
   "source": [
    " Since there are multiple such taxa, we need to decide which one is the reference we should use in going forward. For this, since the OMA RDF does not store GO annotations, we will federate with UniProt to compute the number of annotations per taxon. We will consider the reference one to be the highest annotated one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f636a92",
   "metadata": {},
   "source": [
    "### 2. Federate with UniProt to identify OMA Yeast taxon with most GO term annotations (this will be the reference one)\n",
    "Note: despite how few taxa are transmitted from OMA to UniProt, this query takes very long to execute. This is  because all annotations need to be fetched from uniprot and sent back over before computing the aggregation (computing aggregations on the UniProt side fails - we have a similar example using Bgee below)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "31868a3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 56.6 ms, sys: 14.9 ms, total: 71.5 ms\n",
      "Wall time: 5min 37s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>num_annot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/559292</td>\n",
       "      <td>6417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764097</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764099</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764101</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     taxon num_annot\n",
       "0  http://purl.uniprot.org/taxonomy/559292  6417    \n",
       "1  http://purl.uniprot.org/taxonomy/764097  81      \n",
       "2  http://purl.uniprot.org/taxonomy/764099  70      \n",
       "3  http://purl.uniprot.org/taxonomy/764101  60      "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "query_OMA_UniProt_s_cerevisiae = \"\"\"\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "\n",
    "select ?taxon (count (distinct ?goTerm) as ?num_annot) where {\n",
    "    ?taxon  a up:Taxon.\n",
    "    ?taxon up:scientificName ?name.\n",
    "    filter (strstarts (lcase(?name), \"saccharomyces cerevisiae\" ))\n",
    "    SERVICE <http://sparql.uniprot.org/sparql> {\n",
    "        Select ?taxon ?goTerm where {\n",
    "            ?protein a up:Protein .\n",
    "            ?protein up:classifiedWith ?goTerm .\n",
    "            ?protein  up:organism ?taxon .\n",
    "        } \n",
    "    }\n",
    "} group by ?taxon order by desc(?num_annot) # limit 1 -- if we only want to directly fetch the top one\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "sparql_endpoint_OMA.setQuery(query_OMA_UniProt_s_cerevisiae)\n",
    "sparql_endpoint_OMA.setReturnFormat(JSON)\n",
    "\n",
    "results = sparql_endpoint_OMA.query().convert()\n",
    "\n",
    "pretty_print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac297e63",
   "metadata": {},
   "source": [
    "### 2b. Execute same query at UniProt site (invoking OMA remotely)\n",
    "\n",
    "For performance comparison, we also execute the same query on the UniProt side directly, where we can then perform aggregations locally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "76e183ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 17.4 ms, sys: 5.18 ms, total: 22.6 ms\n",
      "Wall time: 1.28 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_annot</th>\n",
       "      <th>taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6417</td>\n",
       "      <td>http://purl.uniprot.org/taxonomy/559292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>81</td>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>70</td>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>60</td>\n",
       "      <td>http://purl.uniprot.org/taxonomy/764101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  num_annot                                    taxon\n",
       "0  6417      http://purl.uniprot.org/taxonomy/559292\n",
       "1  81        http://purl.uniprot.org/taxonomy/764097\n",
       "2  70        http://purl.uniprot.org/taxonomy/764099\n",
       "3  60        http://purl.uniprot.org/taxonomy/764101"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "query_UniProt_OMA = \"\"\"\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "\n",
    "SELECT ?taxon (count (distinct ?goTerm) as ?num_annot) where {\n",
    "  SERVICE<https://sparql.omabrowser.org/sparql>{\n",
    "select ?taxon where {\n",
    "?taxon  a up:Taxon.\n",
    "?taxon up:scientificName ?name.\n",
    "filter (strstarts (lcase(?name), \"saccharomyces cerevisiae\" ))\n",
    "    } }\n",
    "?protein a up:Protein .\n",
    "    \t?protein up:classifiedWith ?goTerm .\n",
    "    \t?protein  up:organism ?taxon . \n",
    "  \n",
    "} group by ?taxon order by desc(?num_annot) # limit 1 -- if we only want to directly fetch the top one\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "sparql_endpoint_UniProt.setQuery(query_UniProt_OMA)\n",
    "sparql_endpoint_UniProt.setReturnFormat(JSON)\n",
    "\n",
    "results = sparql_endpoint_UniProt.query().convert()\n",
    "\n",
    "pretty_print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37cc1b54",
   "metadata": {},
   "source": [
    "The previous query illustrates that the taxon we are looking for is certainly NCBI ID 559292. We will use this throughout the remainder of the tutorial directly.\n",
    "\n",
    "The difference in execution time between the first variant (5m37s) and the second one (1.2s) is significant.\n",
    "\n",
    "###### **LESSON LEARNED no. 1: generating and transfering large intermediate results can make order of magnitude difference in performance. However, it is not always easy to predict which subqueries will lead to this problem.**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4356772",
   "metadata": {},
   "source": [
    "### 3. Compute orthologs of S. cerevisiae in reference species. \n",
    "\n",
    "Here we just illustrate the OMA query required to compute orthologs of S. cerevisiae:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3d1a9781",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_OMA_orthologs = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dct: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX void: <http://rdfs.org/ns/void#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "\n",
    "select ?p ?species ?orthologGeneEns ?anat  where {\n",
    "?p a orth:Protein.\n",
    "?p orth:organism/obo:RO_0002162 taxon:559292.\n",
    "?cluster a orth:OrthologsCluster.\n",
    "?cluster orth:hasHomologousMember ?node1.\n",
    "?cluster orth:hasHomologousMember ?node2. \n",
    "?node2 orth:hasHomologousMember* ?ortholog. \n",
    "?node1 orth:hasHomologousMember* ?p.\n",
    "?ortholog sio:SIO_010079 ?gene . #is encoded by\n",
    "?gene lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\n",
    "  \n",
    "filter(?node1 != ?node2) \n",
    "}\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c51918d8",
   "metadata": {},
   "source": [
    "### 4. Federate OMA and Bgee to compute orthologs and retrieve anatomical entities (tissues/organs) where these orthologs are highly expressed\n",
    "\n",
    "In principle, the following query should be used. However, this will likely fail because the result size is too big."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "20d63c1c",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1105\u001b[0m             \u001b[0;34m:\u001b[0m\u001b[0mrtype\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;32mclass\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mQueryResult\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0minstance\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \"\"\"\n\u001b[0;32m-> 1107\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mQueryResult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1108\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1109\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mqueryAndConvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1071\u001b[0m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murlopener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1072\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1073\u001b[0;31m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murlopener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1074\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturnFormat\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1075\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0murllib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mHTTPError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    212\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    213\u001b[0m         \u001b[0mopener\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_opener\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 214\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mopener\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    215\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    216\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0minstall_opener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopener\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    521\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mprocessor\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprocess_response\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotocol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    522\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 523\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    524\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    525\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    630\u001b[0m         \u001b[0;31m# request was successfully received, understood, and accepted.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    631\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 632\u001b[0;31m             response = self.parent.error(\n\u001b[0m\u001b[1;32m    633\u001b[0m                 'http', request, response, code, msg, hdrs)\n\u001b[1;32m    634\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    553\u001b[0m             \u001b[0mhttp_err\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    554\u001b[0m         \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 555\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    556\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    557\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    492\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    493\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 494\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    495\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    496\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_302\u001b[0;34m(self, req, fp, code, msg, headers)\u001b[0m\n\u001b[1;32m    745\u001b[0m         \u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    746\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 747\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    749\u001b[0m     \u001b[0mhttp_error_301\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhttp_error_303\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhttp_error_307\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhttp_error_302\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    515\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    516\u001b[0m         \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maudit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'urllib.Request'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_url\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_method\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 517\u001b[0;31m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_open\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    518\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    519\u001b[0m         \u001b[0;31m# post-process response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36m_open\u001b[0;34m(self, req, data)\u001b[0m\n\u001b[1;32m    532\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    533\u001b[0m         \u001b[0mprotocol\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 534\u001b[0;31m         result = self._call_chain(self.handle_open, protocol, protocol +\n\u001b[0m\u001b[1;32m    535\u001b[0m                                   '_open', req)\n\u001b[1;32m    536\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    492\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    493\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 494\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    495\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    496\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttps_open\u001b[0;34m(self, req)\u001b[0m\n\u001b[1;32m   1387\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1388\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mhttps_open\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1389\u001b[0;31m             return self.do_open(http.client.HTTPSConnection, req,\n\u001b[0m\u001b[1;32m   1390\u001b[0m                 context=self._context, check_hostname=self._check_hostname)\n\u001b[1;32m   1391\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mdo_open\u001b[0;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[1;32m   1348\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m# timeout error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1349\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mURLError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1350\u001b[0;31m             \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mh\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1351\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1352\u001b[0m             \u001b[0mh\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/http/client.py\u001b[0m in \u001b[0;36mgetresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1369\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1370\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1371\u001b[0;31m                 \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbegin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1372\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1373\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/http/client.py\u001b[0m in \u001b[0;36mbegin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    317\u001b[0m         \u001b[0;31m# read until we get a non-100 response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    318\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 319\u001b[0;31m             \u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreason\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    320\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstatus\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mCONTINUE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    321\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/http/client.py\u001b[0m in \u001b[0;36m_read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 280\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_MAXLINE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"iso-8859-1\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    281\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0m_MAXLINE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    282\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mLineTooLong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"status line\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    702\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    703\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 704\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    705\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    706\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/ssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1239\u001b[0m                   \u001b[0;34m\"non-zero flags not allowed in calls to recv_into() on %s\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1240\u001b[0m                   self.__class__)\n\u001b[0;32m-> 1241\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1242\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1243\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbytes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/ssl.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1097\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1098\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mbuffer\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1099\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1100\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1101\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sslobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query_Bgee_OMA = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dct: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX void: <http://rdfs.org/ns/void#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "\n",
    "select ?p ?species ?orthologGeneEns ?anat  where {\n",
    "?p a orth:Protein.\n",
    "?p orth:organism/obo:RO_0002162 taxon:559292.\n",
    "?cluster a orth:OrthologsCluster.\n",
    "?cluster orth:hasHomologousMember ?node1.\n",
    "?cluster orth:hasHomologousMember ?node2. \n",
    "?node2 orth:hasHomologousMember* ?ortholog. \n",
    "?node1 orth:hasHomologousMember* ?p.\n",
    "?ortholog sio:SIO_010079 ?gene . #is encoded by\n",
    "?gene lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\n",
    "  \n",
    "filter(?node1 != ?node2) \n",
    "  \n",
    "  SERVICE <https://bgee.org/sparql/> {\n",
    "\t\t\tSELECT DISTINCT ?orthologGeneEns ?anat {\n",
    "\t\t\t\t?geneB a orth:Gene .\n",
    "\t\t\t\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\n",
    "                ?expr a <http://purl.org/genex#Expression> .\n",
    "                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\n",
    "                ?expr genex:hasExpressionLevel ?exprLevel .\n",
    "                FILTER (?exprLevel > 99) # highly expressed      \n",
    "                ?expr genex:hasExpressionCondition ?cond .\n",
    "                ?cond genex:hasAnatomicalEntity ?anat .\n",
    "\t\t\t}\n",
    "  }\n",
    " \n",
    "}\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "sparql_endpoint_OMA.setQuery(query_Bgee_OMA)\n",
    "sparql_endpoint_OMA.setReturnFormat(JSON)\n",
    "\n",
    "results = sparql_endpoint_OMA.query().convert()\n",
    "\n",
    "pretty_print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb6472e9",
   "metadata": {},
   "source": [
    "### As an optimisation, we want to compute *expression breadth* per ortholog by federating with Bgee and performing the aggregation (counting anatomical entities) on the Bgee side. This will significantly reduce the size of results being sent back over (to a single count per orthologous gene).\n",
    "\n",
    "First (obvious) attempt is to join on the gene ID (here, the Ensembl gene, stored in *?orthologousGeneEns*), which would also be used to group by and count tissues in the Bgee side:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "fbbb99fb",
   "metadata": {},
   "outputs": [
    {
     "ename": "EndPointInternalError",
     "evalue": "EndPointInternalError: endpoint returned code 500 and response. \n\nResponse:\nb\"Virtuoso RDFZZ Error DB.DBA.SPARQL_REXEC('https://bgee.org/sparql/', ...) returned Content-Type 'text/plain' status 'HTTP/1.1 400 Bad Request\\r\\n'\\nVirtuoso 37000 Error SP030: SPARQL compiler, line 12: syntax error at '}'\\n\\nSPARQL query:\\ndefine sql:big-data-const 0  SELECT ?expr_breadth\\n WHERE { \\n      { SELECT  <http://rdf.ebi.ac.uk/resource/ensembl/ENSGACG00000014827>\\n          ( COUNT( DISTINCT ?anat) AS ?expr_breadth)\\n         WHERE {  ?geneB <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://purl.org/net/orth#Gene> ;\\n                 <http://purl.org/lscr#xrefEnsemblGene> <http://rdf.ebi.ac.uk/resource/ensembl/ENSGACG00000014827> . ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB ;\\n                 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://purl.org/genex#Expression> ;\\n                 <http://purl.org/genex#hasConfidenceLevel> <http://purl.obolibrary.org/obo/CIO_0000029> ;\\n                 <http://purl.org/genex#hasExpressionLevel> ?exprLevel ;\\n                 <http://purl.org/genex#hasExpressionCondition> ?cond . ?cond <http://purl.org/genex#hasAnatomicalEntity> ?anat .\\n            FILTER ( ?exprLe\\n\\nSPARQL query:\\ndefine sql:big-data-const 0 \\n#output-format:application/sparql-results+json\\n\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX owl: <http://www.w3.org/2002/07/owl#>\\nPREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\\nPREFIX dc: <http://purl.org/dc/elements/1.1/>\\nPREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX oma: <http://omabrowser.org/ontology/oma#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX void: <http://rdfs.org/ns/void#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSelect ?p ?species ?orthologGeneEns ?expr_breadth where {\\n{\\n  SELECT DISTINCT ?p ?species ?orthologGeneEns where { \\n?p a orth:Protein.\\n?p orth:organism/obo:RO_0002162 taxon:559292.\\n?cluster a orth:OrthologsCluster.\\n?cluster orth:hasHomologousMember ?node1.\\n?cluster orth:hasHomologousMember ?node2. \\n?node2 orth:hasHomologousMember* ?ortholog. \\n?node1 orth:hasHomologousMember* ?p.\\n?ortholog sio:SIO_010079 ?gene . #is encoded by\\n?gene lscr:xrefEnsemblGene ?orthologGeneEns .\\n?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\\n  \\nfilter(?node1 != ?node2) \\n\\n} limit 100 }\\n\\n{\\n  SERVICE <https://bgee.org/sparql/> {\\n\\t\\t\\tSELECT ?orthologGeneEns (count (distinct ?anat) as ?expr_breadth) {\\n\\t\\t\\t\\t?geneB a orth:Gene .\\n\\t\\t\\t\\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\\n                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\\n                ?expr a <http://purl.org/genex#Expression> .\\n                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\\n                ?expr genex:hasExpressionLevel ?exprLevel .\\n                FILTER (?exprLevel > 99) # highly expressed      \\n                ?expr genex:hasExpressionCondition ?cond .\\n                ?cond genex:hasAnatomicalEntity ?anat .\\n\\t\\t\\t}  group by ?orthologGeneEns\\n            }\\n            } \\n}\\n\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mHTTPError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1072\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1073\u001b[0;31m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murlopener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1074\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturnFormat\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    213\u001b[0m         \u001b[0mopener\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_opener\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 214\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mopener\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    215\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    522\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 523\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    524\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    631\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 632\u001b[0;31m             response = self.parent.error(\n\u001b[0m\u001b[1;32m    633\u001b[0m                 'http', request, response, code, msg, hdrs)\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    554\u001b[0m         \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 555\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    556\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    493\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 494\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    495\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_302\u001b[0;34m(self, req, fp, code, msg, headers)\u001b[0m\n\u001b[1;32m    746\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 747\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    522\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 523\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    524\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    631\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 632\u001b[0;31m             response = self.parent.error(\n\u001b[0m\u001b[1;32m    633\u001b[0m                 'http', request, response, code, msg, hdrs)\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    560\u001b[0m             \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'default'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'http_error_default'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0morig_args\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 561\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    562\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    493\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 494\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    495\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_default\u001b[0;34m(self, req, fp, code, msg, hdrs)\u001b[0m\n\u001b[1;32m    640\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mhttp_error_default\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 641\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mHTTPError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_url\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    642\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mHTTPError\u001b[0m: HTTP Error 500: SPARQL Request Failed",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mEndPointInternalError\u001b[0m                     Traceback (most recent call last)",
      "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1105\u001b[0m             \u001b[0;34m:\u001b[0m\u001b[0mrtype\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;32mclass\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mQueryResult\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0minstance\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \"\"\"\n\u001b[0;32m-> 1107\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mQueryResult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1108\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1109\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mqueryAndConvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1083\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mURITooLong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1084\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m500\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1085\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mEndPointInternalError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1086\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1087\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mEndPointInternalError\u001b[0m: EndPointInternalError: endpoint returned code 500 and response. \n\nResponse:\nb\"Virtuoso RDFZZ Error DB.DBA.SPARQL_REXEC('https://bgee.org/sparql/', ...) returned Content-Type 'text/plain' status 'HTTP/1.1 400 Bad Request\\r\\n'\\nVirtuoso 37000 Error SP030: SPARQL compiler, line 12: syntax error at '}'\\n\\nSPARQL query:\\ndefine sql:big-data-const 0  SELECT ?expr_breadth\\n WHERE { \\n      { SELECT  <http://rdf.ebi.ac.uk/resource/ensembl/ENSGACG00000014827>\\n          ( COUNT( DISTINCT ?anat) AS ?expr_breadth)\\n         WHERE {  ?geneB <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://purl.org/net/orth#Gene> ;\\n                 <http://purl.org/lscr#xrefEnsemblGene> <http://rdf.ebi.ac.uk/resource/ensembl/ENSGACG00000014827> . ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB ;\\n                 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://purl.org/genex#Expression> ;\\n                 <http://purl.org/genex#hasConfidenceLevel> <http://purl.obolibrary.org/obo/CIO_0000029> ;\\n                 <http://purl.org/genex#hasExpressionLevel> ?exprLevel ;\\n                 <http://purl.org/genex#hasExpressionCondition> ?cond . ?cond <http://purl.org/genex#hasAnatomicalEntity> ?anat .\\n            FILTER ( ?exprLe\\n\\nSPARQL query:\\ndefine sql:big-data-const 0 \\n#output-format:application/sparql-results+json\\n\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX owl: <http://www.w3.org/2002/07/owl#>\\nPREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\\nPREFIX dc: <http://purl.org/dc/elements/1.1/>\\nPREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX oma: <http://omabrowser.org/ontology/oma#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX void: <http://rdfs.org/ns/void#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSelect ?p ?species ?orthologGeneEns ?expr_breadth where {\\n{\\n  SELECT DISTINCT ?p ?species ?orthologGeneEns where { \\n?p a orth:Protein.\\n?p orth:organism/obo:RO_0002162 taxon:559292.\\n?cluster a orth:OrthologsCluster.\\n?cluster orth:hasHomologousMember ?node1.\\n?cluster orth:hasHomologousMember ?node2. \\n?node2 orth:hasHomologousMember* ?ortholog. \\n?node1 orth:hasHomologousMember* ?p.\\n?ortholog sio:SIO_010079 ?gene . #is encoded by\\n?gene lscr:xrefEnsemblGene ?orthologGeneEns .\\n?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\\n  \\nfilter(?node1 != ?node2) \\n\\n} limit 100 }\\n\\n{\\n  SERVICE <https://bgee.org/sparql/> {\\n\\t\\t\\tSELECT ?orthologGeneEns (count (distinct ?anat) as ?expr_breadth) {\\n\\t\\t\\t\\t?geneB a orth:Gene .\\n\\t\\t\\t\\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\\n                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\\n                ?expr a <http://purl.org/genex#Expression> .\\n                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\\n                ?expr genex:hasExpressionLevel ?exprLevel .\\n                FILTER (?exprLevel > 99) # highly expressed      \\n                ?expr genex:hasExpressionCondition ?cond .\\n                ?cond genex:hasAnatomicalEntity ?anat .\\n\\t\\t\\t}  group by ?orthologGeneEns\\n            }\\n            } \\n}\\n\""
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query_limits = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dct: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX void: <http://rdfs.org/ns/void#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "\n",
    "Select ?p ?species ?orthologGeneEns ?expr_breadth where {\n",
    "{\n",
    "  SELECT DISTINCT ?p ?species ?orthologGeneEns where { \n",
    "?p a orth:Protein.\n",
    "?p orth:organism/obo:RO_0002162 taxon:559292.\n",
    "?cluster a orth:OrthologsCluster.\n",
    "?cluster orth:hasHomologousMember ?node1.\n",
    "?cluster orth:hasHomologousMember ?node2. \n",
    "?node2 orth:hasHomologousMember* ?ortholog. \n",
    "?node1 orth:hasHomologousMember* ?p.\n",
    "?ortholog sio:SIO_010079 ?gene . #is encoded by\n",
    "?gene lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\n",
    "  \n",
    "filter(?node1 != ?node2) \n",
    "\n",
    "} limit 100 }\n",
    "\n",
    "{\n",
    "  SERVICE <https://bgee.org/sparql/> {\n",
    "\t\t\tSELECT ?orthologGeneEns (count (distinct ?anat) as ?expr_breadth) {\n",
    "\t\t\t\t?geneB a orth:Gene .\n",
    "\t\t\t\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\n",
    "                ?expr a <http://purl.org/genex#Expression> .\n",
    "                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\n",
    "                ?expr genex:hasExpressionLevel ?exprLevel .\n",
    "                FILTER (?exprLevel > 99) # highly expressed      \n",
    "                ?expr genex:hasExpressionCondition ?cond .\n",
    "                ?cond genex:hasAnatomicalEntity ?anat .\n",
    "\t\t\t}  group by ?orthologGeneEns\n",
    "            }\n",
    "            } \n",
    "}\n",
    "\"\"\"\n",
    "sparql_endpoint_OMA.setQuery(query_limits)\n",
    "sparql_endpoint_OMA.setReturnFormat(JSON)\n",
    "\n",
    "\n",
    "results_groupby_Bgee = sparql_endpoint_OMA.query().convert()\n",
    "#print(results)\n",
    "#pretty_print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b28a25cc",
   "metadata": {},
   "source": [
    "This initial attempt fails, because at query time, the values of *?orthologousGeneEns* are sent 1 by 1 from OMA to Bgee, hence there is no option to group by this anymore. In the error message above, you can see that in the Bgee service invocation the variable is already replaced by the URI <http://rdf.ebi.ac.uk/resource/ensembl/ENSGACG00000014827>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7638aa4",
   "metadata": {},
   "source": [
    "### Second attempt: removing the groupby and simply counting elements in the remote query invocation (since the variable is replaced at query time already). \n",
    "\n",
    "This also fails, this time with a more cryptic Virtuoso error (Virtuoso HTCLI Error HC001: Read Error in HTTP Client):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "ce66ca56",
   "metadata": {},
   "outputs": [
    {
     "ename": "EndPointInternalError",
     "evalue": "EndPointInternalError: endpoint returned code 500 and response. \n\nResponse:\nb'Virtuoso HTCLI Error HC001: Read Error in HTTP Client\\n\\nSPARQL query:\\ndefine sql:big-data-const 0 \\n#output-format:application/sparql-results+json\\n\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX owl: <http://www.w3.org/2002/07/owl#>\\nPREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\\nPREFIX dc: <http://purl.org/dc/elements/1.1/>\\nPREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX oma: <http://omabrowser.org/ontology/oma#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX void: <http://rdfs.org/ns/void#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSelect ?p ?species ?orthologGeneEns ?expr_breadth where {\\n{\\n  SELECT DISTINCT ?p ?species ?orthologGeneEns where { \\n?p a orth:Protein.\\n?p orth:organism/obo:RO_0002162 taxon:559292.\\n?cluster a orth:OrthologsCluster.\\n?cluster orth:hasHomologousMember ?node1.\\n?cluster orth:hasHomologousMember ?node2. \\n?node2 orth:hasHomologousMember* ?ortholog. \\n?node1 orth:hasHomologousMember* ?p.\\n?ortholog sio:SIO_010079 ?gene . #is encoded by\\n?gene lscr:xrefEnsemblGene ?orthologGeneEns .\\n?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\\n  \\nfilter(?node1 != ?node2) \\n\\n} limit 100 }\\n\\n{\\n  SERVICE <https://bgee.org/sparql/> {\\n\\t\\t\\tSELECT (count (distinct ?anat) as ?expr_breadth) {\\n\\t\\t\\t\\t?geneB a orth:Gene .\\n\\t\\t\\t\\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\\n                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\\n                ?expr a <http://purl.org/genex#Expression> .\\n                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\\n                ?expr genex:hasExpressionLevel ?exprLevel .\\n                FILTER (?exprLevel > 99) # highly expressed      \\n                ?expr genex:hasExpressionCondition ?cond .\\n                ?cond genex:hasAnatomicalEntity ?anat .\\n\\t\\t\\t} \\n            }\\n            } \\n}\\n'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mHTTPError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1072\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1073\u001b[0;31m                 \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0murlopener\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1074\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturnFormat\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[1;32m    213\u001b[0m         \u001b[0mopener\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_opener\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 214\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mopener\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    215\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    522\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 523\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    524\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    631\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 632\u001b[0;31m             response = self.parent.error(\n\u001b[0m\u001b[1;32m    633\u001b[0m                 'http', request, response, code, msg, hdrs)\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    554\u001b[0m         \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 555\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    556\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    493\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 494\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    495\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_302\u001b[0;34m(self, req, fp, code, msg, headers)\u001b[0m\n\u001b[1;32m    746\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 747\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    748\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    522\u001b[0m             \u001b[0mmeth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessor\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 523\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    524\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_response\u001b[0;34m(self, request, response)\u001b[0m\n\u001b[1;32m    631\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m200\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcode\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m300\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 632\u001b[0;31m             response = self.parent.error(\n\u001b[0m\u001b[1;32m    633\u001b[0m                 'http', request, response, code, msg, hdrs)\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36merror\u001b[0;34m(self, proto, *args)\u001b[0m\n\u001b[1;32m    560\u001b[0m             \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'default'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'http_error_default'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0morig_args\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 561\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_chain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    562\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    493\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 494\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    495\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/urllib/request.py\u001b[0m in \u001b[0;36mhttp_error_default\u001b[0;34m(self, req, fp, code, msg, hdrs)\u001b[0m\n\u001b[1;32m    640\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mhttp_error_default\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 641\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mHTTPError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfull_url\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhdrs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    642\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mHTTPError\u001b[0m: HTTP Error 500: SPARQL Request Failed",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mEndPointInternalError\u001b[0m                     Traceback (most recent call last)",
      "\u001b[0;32m<timed exec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1105\u001b[0m             \u001b[0;34m:\u001b[0m\u001b[0mrtype\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;32mclass\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mQueryResult\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0minstance\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1106\u001b[0m         \"\"\"\n\u001b[0;32m-> 1107\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mQueryResult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1108\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1109\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mqueryAndConvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/SPARQLWrapper/Wrapper.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1083\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mURITooLong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1084\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m500\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1085\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mEndPointInternalError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1086\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1087\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mEndPointInternalError\u001b[0m: EndPointInternalError: endpoint returned code 500 and response. \n\nResponse:\nb'Virtuoso HTCLI Error HC001: Read Error in HTTP Client\\n\\nSPARQL query:\\ndefine sql:big-data-const 0 \\n#output-format:application/sparql-results+json\\n\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX owl: <http://www.w3.org/2002/07/owl#>\\nPREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\\nPREFIX dc: <http://purl.org/dc/elements/1.1/>\\nPREFIX dct: <http://purl.org/dc/terms/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nPREFIX oma: <http://omabrowser.org/ontology/oma#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX void: <http://rdfs.org/ns/void#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSelect ?p ?species ?orthologGeneEns ?expr_breadth where {\\n{\\n  SELECT DISTINCT ?p ?species ?orthologGeneEns where { \\n?p a orth:Protein.\\n?p orth:organism/obo:RO_0002162 taxon:559292.\\n?cluster a orth:OrthologsCluster.\\n?cluster orth:hasHomologousMember ?node1.\\n?cluster orth:hasHomologousMember ?node2. \\n?node2 orth:hasHomologousMember* ?ortholog. \\n?node1 orth:hasHomologousMember* ?p.\\n?ortholog sio:SIO_010079 ?gene . #is encoded by\\n?gene lscr:xrefEnsemblGene ?orthologGeneEns .\\n?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\\n  \\nfilter(?node1 != ?node2) \\n\\n} limit 100 }\\n\\n{\\n  SERVICE <https://bgee.org/sparql/> {\\n\\t\\t\\tSELECT (count (distinct ?anat) as ?expr_breadth) {\\n\\t\\t\\t\\t?geneB a orth:Gene .\\n\\t\\t\\t\\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\\n                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\\n                ?expr a <http://purl.org/genex#Expression> .\\n                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\\n                ?expr genex:hasExpressionLevel ?exprLevel .\\n                FILTER (?exprLevel > 99) # highly expressed      \\n                ?expr genex:hasExpressionCondition ?cond .\\n                ?cond genex:hasAnatomicalEntity ?anat .\\n\\t\\t\\t} \\n            }\\n            } \\n}\\n'"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query_limits = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dct: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX void: <http://rdfs.org/ns/void#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "\n",
    "Select ?p ?species ?orthologGeneEns ?expr_breadth where {\n",
    "{\n",
    "  SELECT DISTINCT ?p ?species ?orthologGeneEns where { \n",
    "?p a orth:Protein.\n",
    "?p orth:organism/obo:RO_0002162 taxon:559292.\n",
    "?cluster a orth:OrthologsCluster.\n",
    "?cluster orth:hasHomologousMember ?node1.\n",
    "?cluster orth:hasHomologousMember ?node2. \n",
    "?node2 orth:hasHomologousMember* ?ortholog. \n",
    "?node1 orth:hasHomologousMember* ?p.\n",
    "?ortholog sio:SIO_010079 ?gene . #is encoded by\n",
    "?gene lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\n",
    "  \n",
    "filter(?node1 != ?node2) \n",
    "\n",
    "} limit 100 }\n",
    "\n",
    "{\n",
    "  SERVICE <https://bgee.org/sparql/> {\n",
    "\t\t\tSELECT (count (distinct ?anat) as ?expr_breadth) {\n",
    "\t\t\t\t?geneB a orth:Gene .\n",
    "\t\t\t\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\n",
    "                ?expr a <http://purl.org/genex#Expression> .\n",
    "                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\n",
    "                ?expr genex:hasExpressionLevel ?exprLevel .\n",
    "                FILTER (?exprLevel > 99) # highly expressed      \n",
    "                ?expr genex:hasExpressionCondition ?cond .\n",
    "                ?cond genex:hasAnatomicalEntity ?anat .\n",
    "\t\t\t} \n",
    "            }\n",
    "            } \n",
    "}\n",
    "\"\"\"\n",
    "sparql_endpoint_OMA.setQuery(query_limits)\n",
    "sparql_endpoint_OMA.setReturnFormat(JSON)\n",
    "\n",
    "\n",
    "results_no_groupby_Bgee = sparql_endpoint_OMA.query().convert()\n",
    "#print(results)\n",
    "#pretty_print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55a1aef4",
   "metadata": {},
   "source": [
    "### Attempted aggregation via SPARQL - failed....\n",
    "\n",
    "###### **LESSON LEARNED no. 2: writing federated queries with aggregations is extremely challenging**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e392c380",
   "metadata": {},
   "source": [
    "Given these errors, we abandon the obvious path of aggregating on the remote side and choose the inefficient (but feasible) path of retrieving data from the remote server and aggregating it (i.e. counting) locally. We cannot do this for the entirety of the orthologous genes as it would lead to a too large results set and likely timeouts, therefore we have to also limit the results in the first part of the query to have a chance of illustrating how this works:\n",
    "\n",
    "We limit the results from OMA to the top 30K (protein, orthologous) pairs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "5a1f8353",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.08 s, sys: 1.24 s, total: 2.32 s\n",
      "Wall time: 51min 52s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "query_limits = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dct: <http://purl.org/dc/terms/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX oma: <http://omabrowser.org/ontology/oma#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX void: <http://rdfs.org/ns/void#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "\n",
    "Select ?p ?species ?orthologGeneEns ?anat where {\n",
    "{\n",
    "  SELECT DISTINCT ?p ?species ?orthologGeneEns where { \n",
    "?p a orth:Protein.\n",
    "?p orth:organism/obo:RO_0002162 taxon:559292.\n",
    "?cluster a orth:OrthologsCluster.\n",
    "?cluster orth:hasHomologousMember ?node1.\n",
    "?cluster orth:hasHomologousMember ?node2. \n",
    "?node2 orth:hasHomologousMember* ?ortholog. \n",
    "?node1 orth:hasHomologousMember* ?p.\n",
    "?ortholog sio:SIO_010079 ?gene . #is encoded by\n",
    "?gene lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "?ortholog orth:organism/obo:RO_0002162/up:scientificName ?species.\n",
    "  \n",
    "filter(?node1 != ?node2) \n",
    "\n",
    "} limit 30000 }\n",
    "\n",
    "{\n",
    "  SERVICE <https://bgee.org/sparql/> {\n",
    "\t\t\tSELECT ?orthologGeneEns ?anat {\n",
    "\t\t\t\t?geneB a orth:Gene .\n",
    "\t\t\t\t?geneB lscr:xrefEnsemblGene ?orthologGeneEns .\n",
    "                ?expr <http://purl.org/genex#hasSequenceUnit> ?geneB.\n",
    "                ?expr a <http://purl.org/genex#Expression> .\n",
    "                ?expr genex:hasConfidenceLevel obo:CIO_0000029 . # high confidence level\n",
    "                ?expr genex:hasExpressionLevel ?exprLevel .\n",
    "                FILTER (?exprLevel > 99) # highly expressed      \n",
    "                ?expr genex:hasExpressionCondition ?cond .\n",
    "                ?cond genex:hasAnatomicalEntity ?anat .\n",
    "\t\t\t}  \n",
    "            }\n",
    "            } \n",
    "}\n",
    "\"\"\"\n",
    "sparql_endpoint_OMA.setQuery(query_limits)\n",
    "sparql_endpoint_OMA.setReturnFormat(JSON)\n",
    "\n",
    "\n",
    "results = sparql_endpoint_OMA.query().convert()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "d8ffa0b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p</th>\n",
       "      <th>species</th>\n",
       "      <th>orthologGeneEns</th>\n",
       "      <th>anat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GO_0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GO_0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>Oryzias latipes</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175511</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GO_0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175512</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0000079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175513</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GO_0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175514</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0003101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175515</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>Monodelphis domestica</td>\n",
       "      <td>http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432</td>\n",
       "      <td>http://purl.obolibrary.org/obo/GO_0005575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175516 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 p                species  \\\n",
       "0       https://omabrowser.org/oma/info/YEAST03651  Oryzias latipes         \n",
       "1       https://omabrowser.org/oma/info/YEAST03651  Oryzias latipes         \n",
       "2       https://omabrowser.org/oma/info/YEAST03651  Oryzias latipes         \n",
       "3       https://omabrowser.org/oma/info/YEAST03651  Oryzias latipes         \n",
       "4       https://omabrowser.org/oma/info/YEAST03651  Oryzias latipes         \n",
       "...                                            ...              ...         \n",
       "175511  https://omabrowser.org/oma/info/YEAST03807  Monodelphis domestica   \n",
       "175512  https://omabrowser.org/oma/info/YEAST03807  Monodelphis domestica   \n",
       "175513  https://omabrowser.org/oma/info/YEAST03807  Monodelphis domestica   \n",
       "175514  https://omabrowser.org/oma/info/YEAST03807  Monodelphis domestica   \n",
       "175515  https://omabrowser.org/oma/info/YEAST03807  Monodelphis domestica   \n",
       "\n",
       "                                                 orthologGeneEns  \\\n",
       "0       http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367   \n",
       "1       http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367   \n",
       "2       http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367   \n",
       "3       http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367   \n",
       "4       http://rdf.ebi.ac.uk/resource/ensembl/ENSORLG00000018367   \n",
       "...                                                          ...   \n",
       "175511  http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432   \n",
       "175512  http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432   \n",
       "175513  http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432   \n",
       "175514  http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432   \n",
       "175515  http://rdf.ebi.ac.uk/resource/ensembl/ENSMODG00000017432   \n",
       "\n",
       "                                                 anat  \n",
       "0       http://purl.obolibrary.org/obo/UBERON_0000468  \n",
       "1       http://purl.obolibrary.org/obo/GO_0005575      \n",
       "2       http://purl.obolibrary.org/obo/UBERON_0000465  \n",
       "3       http://purl.obolibrary.org/obo/GO_0005575      \n",
       "4       http://purl.obolibrary.org/obo/UBERON_0000061  \n",
       "...                                               ...  \n",
       "175511  http://purl.obolibrary.org/obo/GO_0005575      \n",
       "175512  http://purl.obolibrary.org/obo/UBERON_0000079  \n",
       "175513  http://purl.obolibrary.org/obo/GO_0005575      \n",
       "175514  http://purl.obolibrary.org/obo/UBERON_0003101  \n",
       "175515  http://purl.obolibrary.org/obo/GO_0005575      \n",
       "\n",
       "[175516 rows x 4 columns]"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pretty_print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cfa2de9",
   "metadata": {},
   "source": [
    "This query completed in 51m and produced 175K rows."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "666235f3",
   "metadata": {},
   "source": [
    "Now let us explore a bit the results. In particular, let's count for how many proteins we have data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "9f8574c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35\n",
      "507\n"
     ]
    }
   ],
   "source": [
    "proteins = [entry[\"p\"][\"value\"] for entry in results[\"results\"][\"bindings\"]]\n",
    "print(len(set(proteins)))\n",
    "\n",
    "orthologs = [entry[\"orthologGeneEns\"][\"value\"] for entry in results[\"results\"][\"bindings\"]]\n",
    "print(len(set(orthologs)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c62f52f5",
   "metadata": {},
   "source": [
    "### Compute aggregation directly in Python\n",
    "\n",
    "Assuming we would have retrieved in the \"results\" variable all orthologs of S. cerevisiae in reference species contained in Bgee and the corresponding tissues where these orthologs are highly expressed, we could perform the subsequent aggregations offline (computing the number of orthologs per protein and the average expression breadth of these orthologs). Here we illustrate how this could be achieved, although results were truncated, therefore we will have only a partial view of the data.\n",
    "\n",
    "In *results*, *p* is the protein in S. cerevisiae, *orthologGeneEns* is the orthologous protein (found in *species*) and *anat* is the tissue where the orthologous protein is highly expressed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "6e81a1d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create DataFrame with {Protein, Ortholog} pairs\n",
    "\n",
    "# in the SPARQL results, data is retrieved in the \"value\" field of each entry in the \"bindings\" of the results\n",
    "proteins_orthologs = [{'Protein_S_cerevisiae': entry[\"p\"][\"value\"], 'Orthologous_Protein': entry[\"orthologGeneEns\"][\"value\"]} for \n",
    "                      entry in results[\"results\"][\"bindings\"]]\n",
    "\n",
    "orthologs_df = pd.DataFrame.from_records(proteins_orthologs, columns=['Protein_S_cerevisiae', 'Orthologous_Protein'])\n",
    "orthologs_df = orthologs_df.drop_duplicates()\n",
    "orthologs_df[\"num_orthologs\"] = orthologs_df.groupby(\"Protein_S_cerevisiae\")[\"Orthologous_Protein\"].transform(\"count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "id": "f83dee9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_orthologs_per_protein_sorted = orthologs_df[[\"Protein_S_cerevisiae\",\"num_orthologs\"]].drop_duplicates().sort_values(\"num_orthologs\", ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "62290a34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein_S_cerevisiae</th>\n",
       "      <th>num_orthologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23138</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03625</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92428</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03653</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03641</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61174</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03685</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144892</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03626</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70778</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03708</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16866</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03687</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47380</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03697</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90810</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03706</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167740</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03675</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170320</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03845</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175348</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15020</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03700</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92138</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03712</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174478</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03766</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166210</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03722</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90738</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03730</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166958</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03703</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47216</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03727</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15316</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03693</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47320</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03701</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16674</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03669</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167254</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03682</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03665</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03656</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167728</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03695</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03743</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166928</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03670</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90794</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03690</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16822</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03714</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92060</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03716</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92048</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03691</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70776</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03739</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166900</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03696</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Protein_S_cerevisiae  num_orthologs\n",
       "23138   https://omabrowser.org/oma/info/YEAST03625  169          \n",
       "92428   https://omabrowser.org/oma/info/YEAST03653  45           \n",
       "396     https://omabrowser.org/oma/info/YEAST03641  40           \n",
       "61174   https://omabrowser.org/oma/info/YEAST03685  40           \n",
       "144892  https://omabrowser.org/oma/info/YEAST03626  34           \n",
       "70778   https://omabrowser.org/oma/info/YEAST03708  30           \n",
       "16866   https://omabrowser.org/oma/info/YEAST03687  29           \n",
       "47380   https://omabrowser.org/oma/info/YEAST03697  29           \n",
       "90810   https://omabrowser.org/oma/info/YEAST03706  15           \n",
       "167740  https://omabrowser.org/oma/info/YEAST03675  14           \n",
       "170320  https://omabrowser.org/oma/info/YEAST03845  10           \n",
       "175348  https://omabrowser.org/oma/info/YEAST03807  6            \n",
       "15020   https://omabrowser.org/oma/info/YEAST03700  6            \n",
       "92138   https://omabrowser.org/oma/info/YEAST03712  6            \n",
       "174478  https://omabrowser.org/oma/info/YEAST03766  4            \n",
       "166210  https://omabrowser.org/oma/info/YEAST03722  4            \n",
       "0       https://omabrowser.org/oma/info/YEAST03651  4            \n",
       "90738   https://omabrowser.org/oma/info/YEAST03730  3            \n",
       "166958  https://omabrowser.org/oma/info/YEAST03703  3            \n",
       "47216   https://omabrowser.org/oma/info/YEAST03727  3            \n",
       "15316   https://omabrowser.org/oma/info/YEAST03693  2            \n",
       "47320   https://omabrowser.org/oma/info/YEAST03701  2            \n",
       "16674   https://omabrowser.org/oma/info/YEAST03669  2            \n",
       "167254  https://omabrowser.org/oma/info/YEAST03682  2            \n",
       "350     https://omabrowser.org/oma/info/YEAST03665  1            \n",
       "380     https://omabrowser.org/oma/info/YEAST03656  1            \n",
       "167728  https://omabrowser.org/oma/info/YEAST03695  1            \n",
       "12      https://omabrowser.org/oma/info/YEAST03743  1            \n",
       "166928  https://omabrowser.org/oma/info/YEAST03670  1            \n",
       "90794   https://omabrowser.org/oma/info/YEAST03690  1            \n",
       "16822   https://omabrowser.org/oma/info/YEAST03714  1            \n",
       "92060   https://omabrowser.org/oma/info/YEAST03716  1            \n",
       "92048   https://omabrowser.org/oma/info/YEAST03691  1            \n",
       "70776   https://omabrowser.org/oma/info/YEAST03739  1            \n",
       "166900  https://omabrowser.org/oma/info/YEAST03696  1            "
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_orthologs_per_protein_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c371d226",
   "metadata": {},
   "source": [
    "#### In how many species does each protein have orthologs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "b8c20222",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein_S_cerevisiae</th>\n",
       "      <th>num_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>92428</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03653</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144892</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03626</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03641</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70778</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03708</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61174</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03685</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23138</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03625</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16866</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03687</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47380</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03697</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167740</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03675</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90810</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03706</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170320</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03845</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15020</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03700</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92138</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03712</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166210</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03722</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175348</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90738</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03730</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47216</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03727</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166958</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03703</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174478</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03766</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167254</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03682</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16674</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03669</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15316</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03693</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47320</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03701</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16822</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03714</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92060</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03716</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03743</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166900</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03696</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166928</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03670</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92048</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03691</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167728</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03695</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03656</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03665</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90794</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03690</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70776</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03739</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Protein_S_cerevisiae  num_species\n",
       "92428   https://omabrowser.org/oma/info/YEAST03653  40         \n",
       "144892  https://omabrowser.org/oma/info/YEAST03626  33         \n",
       "396     https://omabrowser.org/oma/info/YEAST03641  29         \n",
       "70778   https://omabrowser.org/oma/info/YEAST03708  28         \n",
       "61174   https://omabrowser.org/oma/info/YEAST03685  28         \n",
       "23138   https://omabrowser.org/oma/info/YEAST03625  27         \n",
       "16866   https://omabrowser.org/oma/info/YEAST03687  26         \n",
       "47380   https://omabrowser.org/oma/info/YEAST03697  26         \n",
       "167740  https://omabrowser.org/oma/info/YEAST03675  14         \n",
       "90810   https://omabrowser.org/oma/info/YEAST03706  13         \n",
       "170320  https://omabrowser.org/oma/info/YEAST03845  10         \n",
       "15020   https://omabrowser.org/oma/info/YEAST03700  5          \n",
       "92138   https://omabrowser.org/oma/info/YEAST03712  5          \n",
       "0       https://omabrowser.org/oma/info/YEAST03651  4          \n",
       "166210  https://omabrowser.org/oma/info/YEAST03722  4          \n",
       "175348  https://omabrowser.org/oma/info/YEAST03807  4          \n",
       "90738   https://omabrowser.org/oma/info/YEAST03730  3          \n",
       "47216   https://omabrowser.org/oma/info/YEAST03727  3          \n",
       "166958  https://omabrowser.org/oma/info/YEAST03703  3          \n",
       "174478  https://omabrowser.org/oma/info/YEAST03766  2          \n",
       "167254  https://omabrowser.org/oma/info/YEAST03682  2          \n",
       "16674   https://omabrowser.org/oma/info/YEAST03669  2          \n",
       "15316   https://omabrowser.org/oma/info/YEAST03693  2          \n",
       "47320   https://omabrowser.org/oma/info/YEAST03701  2          \n",
       "16822   https://omabrowser.org/oma/info/YEAST03714  1          \n",
       "92060   https://omabrowser.org/oma/info/YEAST03716  1          \n",
       "12      https://omabrowser.org/oma/info/YEAST03743  1          \n",
       "166900  https://omabrowser.org/oma/info/YEAST03696  1          \n",
       "166928  https://omabrowser.org/oma/info/YEAST03670  1          \n",
       "92048   https://omabrowser.org/oma/info/YEAST03691  1          \n",
       "167728  https://omabrowser.org/oma/info/YEAST03695  1          \n",
       "380     https://omabrowser.org/oma/info/YEAST03656  1          \n",
       "350     https://omabrowser.org/oma/info/YEAST03665  1          \n",
       "90794   https://omabrowser.org/oma/info/YEAST03690  1          \n",
       "70776   https://omabrowser.org/oma/info/YEAST03739  1          "
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create DataFrame with {Protein, Species} pairs\n",
    "\n",
    "# in the SPARQL results, data is retrieved in the \"value\" field of each entry in the \"bindings\" of the results\n",
    "proteins_orthologs = [{'Protein_S_cerevisiae': entry[\"p\"][\"value\"], 'Species_With_Orthologs': entry[\"species\"][\"value\"]} for \n",
    "                      entry in results[\"results\"][\"bindings\"]]\n",
    "\n",
    "orthologs_species_df = pd.DataFrame.from_records(proteins_orthologs, columns=['Protein_S_cerevisiae', 'Species_With_Orthologs'])\n",
    "orthologs_species_df = orthologs_species_df.drop_duplicates()\n",
    "\n",
    "orthologs_species_df[\"num_species\"] = orthologs_species_df.groupby(\"Protein_S_cerevisiae\")[\"Species_With_Orthologs\"].transform(\"count\")\n",
    "\n",
    "num_species_per_protein_sorted = orthologs_species_df[[\"Protein_S_cerevisiae\",\"num_species\"]].drop_duplicates().sort_values(\"num_species\", ascending = False)\n",
    "\n",
    "num_species_per_protein_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36493b44",
   "metadata": {},
   "source": [
    "Similarly we want to compute the average expression breadth per protein, we can achieve this with the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "d085a080",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create DataFrame with {Protein, Ortholog, Tissue} pairs\n",
    "\n",
    "# in the SPARQL results, data is retrieved in the \"value\" field of each entry in the \"bindings\" of the results\n",
    "proteins_orthologs = [{'Protein_S_cerevisiae': entry[\"p\"][\"value\"], 'Orthologous_Protein': entry[\"orthologGeneEns\"][\"value\"],\n",
    "                       'Tissue_where_Ortholog_expressed': entry[\"anat\"][\"value\"]} for \n",
    "                      entry in results[\"results\"][\"bindings\"]]\n",
    "\n",
    "expr_breadth_df = pd.DataFrame.from_records(proteins_orthologs, columns=['Protein_S_cerevisiae', 'Orthologous_Protein', 'Tissue_where_Ortholog_expressed'])\n",
    "expr_breadth_df = expr_breadth_df.drop_duplicates()\n",
    "# now we want to count number of tissues per ortholog first\n",
    "expr_breadth_df[\"number_of_tissues_per_ortholog\"] = expr_breadth_df.groupby([\"Protein_S_cerevisiae\",'Orthologous_Protein'])[\"Tissue_where_Ortholog_expressed\"].transform(\"count\")\n",
    "\n",
    "# after counting per ortholog, we want to average the expression breadth per protein\n",
    "\n",
    "expr_breadth_df[\"average_expression_breadth\"] = expr_breadth_df.groupby(\"Protein_S_cerevisiae\")[\"number_of_tissues_per_ortholog\"].transform(\"mean\")\n",
    "\n",
    "expr_breadth_df_sorted = expr_breadth_df[[\"Protein_S_cerevisiae\",\"average_expression_breadth\"]].drop_duplicates().sort_values(\"average_expression_breadth\", ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "4f265d8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein_S_cerevisiae</th>\n",
       "      <th>average_expression_breadth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>92428</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03653</td>\n",
       "      <td>194.037491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47380</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03697</td>\n",
       "      <td>166.221399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70778</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03708</td>\n",
       "      <td>132.368049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170320</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03845</td>\n",
       "      <td>132.228986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144892</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03626</td>\n",
       "      <td>124.314199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15316</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03693</td>\n",
       "      <td>121.752577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61174</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03685</td>\n",
       "      <td>119.032086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03641</td>\n",
       "      <td>111.726172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23138</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03625</td>\n",
       "      <td>102.393535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167740</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03675</td>\n",
       "      <td>71.591195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16866</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03687</td>\n",
       "      <td>51.005988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167254</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03682</td>\n",
       "      <td>49.561404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174478</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03766</td>\n",
       "      <td>45.340000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90810</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03706</td>\n",
       "      <td>24.690909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166210</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03722</td>\n",
       "      <td>21.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16822</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03714</td>\n",
       "      <td>20.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166958</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03703</td>\n",
       "      <td>19.681818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92138</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03712</td>\n",
       "      <td>18.605263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16674</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03669</td>\n",
       "      <td>16.083333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03743</td>\n",
       "      <td>15.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15020</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03700</td>\n",
       "      <td>12.228070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47320</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03701</td>\n",
       "      <td>11.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175348</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03807</td>\n",
       "      <td>9.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03656</td>\n",
       "      <td>9.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03651</td>\n",
       "      <td>8.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92060</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03716</td>\n",
       "      <td>8.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90738</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03730</td>\n",
       "      <td>7.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92048</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03691</td>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47216</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03727</td>\n",
       "      <td>6.647059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90794</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03690</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166928</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03670</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03665</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166900</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03696</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167728</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03695</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70776</th>\n",
       "      <td>https://omabrowser.org/oma/info/YEAST03739</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Protein_S_cerevisiae  average_expression_breadth\n",
       "92428   https://omabrowser.org/oma/info/YEAST03653  194.037491                \n",
       "47380   https://omabrowser.org/oma/info/YEAST03697  166.221399                \n",
       "70778   https://omabrowser.org/oma/info/YEAST03708  132.368049                \n",
       "170320  https://omabrowser.org/oma/info/YEAST03845  132.228986                \n",
       "144892  https://omabrowser.org/oma/info/YEAST03626  124.314199                \n",
       "15316   https://omabrowser.org/oma/info/YEAST03693  121.752577                \n",
       "61174   https://omabrowser.org/oma/info/YEAST03685  119.032086                \n",
       "396     https://omabrowser.org/oma/info/YEAST03641  111.726172                \n",
       "23138   https://omabrowser.org/oma/info/YEAST03625  102.393535                \n",
       "167740  https://omabrowser.org/oma/info/YEAST03675  71.591195                 \n",
       "16866   https://omabrowser.org/oma/info/YEAST03687  51.005988                 \n",
       "167254  https://omabrowser.org/oma/info/YEAST03682  49.561404                 \n",
       "174478  https://omabrowser.org/oma/info/YEAST03766  45.340000                 \n",
       "90810   https://omabrowser.org/oma/info/YEAST03706  24.690909                 \n",
       "166210  https://omabrowser.org/oma/info/YEAST03722  21.625000                 \n",
       "16822   https://omabrowser.org/oma/info/YEAST03714  20.000000                 \n",
       "166958  https://omabrowser.org/oma/info/YEAST03703  19.681818                 \n",
       "92138   https://omabrowser.org/oma/info/YEAST03712  18.605263                 \n",
       "16674   https://omabrowser.org/oma/info/YEAST03669  16.083333                 \n",
       "12      https://omabrowser.org/oma/info/YEAST03743  15.000000                 \n",
       "15020   https://omabrowser.org/oma/info/YEAST03700  12.228070                 \n",
       "47320   https://omabrowser.org/oma/info/YEAST03701  11.666667                 \n",
       "175348  https://omabrowser.org/oma/info/YEAST03807  9.266667                  \n",
       "380     https://omabrowser.org/oma/info/YEAST03656  9.000000                  \n",
       "0       https://omabrowser.org/oma/info/YEAST03651  8.500000                  \n",
       "92060   https://omabrowser.org/oma/info/YEAST03716  8.000000                  \n",
       "90738   https://omabrowser.org/oma/info/YEAST03730  7.444444                  \n",
       "92048   https://omabrowser.org/oma/info/YEAST03691  7.000000                  \n",
       "47216   https://omabrowser.org/oma/info/YEAST03727  6.647059                  \n",
       "90794   https://omabrowser.org/oma/info/YEAST03690  6.000000                  \n",
       "166928  https://omabrowser.org/oma/info/YEAST03670  5.000000                  \n",
       "350     https://omabrowser.org/oma/info/YEAST03665  4.000000                  \n",
       "166900  https://omabrowser.org/oma/info/YEAST03696  3.000000                  \n",
       "167728  https://omabrowser.org/oma/info/YEAST03695  3.000000                  \n",
       "70776   https://omabrowser.org/oma/info/YEAST03739  2.000000                  "
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr_breadth_df_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db2d9185",
   "metadata": {},
   "source": [
    "### 6. Intersecting Top N from the two results to predict most likely candidates for essential genes\n",
    "\n",
    "These will be the ones highly conserved across species (orthologs in many species), as well as a high expression breadth of these orthologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "46088258",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select top N from each of the two dataframes above and join them\n",
    "\n",
    "N = 10\n",
    "\n",
    "top_N_expr = expr_breadth_df_sorted.head(N)\n",
    "\n",
    "#top_N_ortho = num_orthologs_per_protein_sorted.head(N)\n",
    "\n",
    "top_N_ortho = num_species_per_protein_sorted.head(N)\n",
    "result = top_N_expr.set_index('Protein_S_cerevisiae').join(top_N_ortho.set_index('Protein_S_cerevisiae'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "8bea87b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>average_expression_breadth</th>\n",
       "      <th>num_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein_S_cerevisiae</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03653</th>\n",
       "      <td>194.037491</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03697</th>\n",
       "      <td>166.221399</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03708</th>\n",
       "      <td>132.368049</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03845</th>\n",
       "      <td>132.228986</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03626</th>\n",
       "      <td>124.314199</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03693</th>\n",
       "      <td>121.752577</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03685</th>\n",
       "      <td>119.032086</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03641</th>\n",
       "      <td>111.726172</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03625</th>\n",
       "      <td>102.393535</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03675</th>\n",
       "      <td>71.591195</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            average_expression_breadth  \\\n",
       "Protein_S_cerevisiae                                                     \n",
       "https://omabrowser.org/oma/info/YEAST03653  194.037491                   \n",
       "https://omabrowser.org/oma/info/YEAST03697  166.221399                   \n",
       "https://omabrowser.org/oma/info/YEAST03708  132.368049                   \n",
       "https://omabrowser.org/oma/info/YEAST03845  132.228986                   \n",
       "https://omabrowser.org/oma/info/YEAST03626  124.314199                   \n",
       "https://omabrowser.org/oma/info/YEAST03693  121.752577                   \n",
       "https://omabrowser.org/oma/info/YEAST03685  119.032086                   \n",
       "https://omabrowser.org/oma/info/YEAST03641  111.726172                   \n",
       "https://omabrowser.org/oma/info/YEAST03625  102.393535                   \n",
       "https://omabrowser.org/oma/info/YEAST03675  71.591195                    \n",
       "\n",
       "                                            num_species  \n",
       "Protein_S_cerevisiae                                     \n",
       "https://omabrowser.org/oma/info/YEAST03653  40.0         \n",
       "https://omabrowser.org/oma/info/YEAST03697  26.0         \n",
       "https://omabrowser.org/oma/info/YEAST03708  28.0         \n",
       "https://omabrowser.org/oma/info/YEAST03845 NaN           \n",
       "https://omabrowser.org/oma/info/YEAST03626  33.0         \n",
       "https://omabrowser.org/oma/info/YEAST03693 NaN           \n",
       "https://omabrowser.org/oma/info/YEAST03685  28.0         \n",
       "https://omabrowser.org/oma/info/YEAST03641  29.0         \n",
       "https://omabrowser.org/oma/info/YEAST03625  27.0         \n",
       "https://omabrowser.org/oma/info/YEAST03675  14.0         "
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f96691c",
   "metadata": {},
   "source": [
    "Finally, we drop NaN rows and get the final candidates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "b1bc51c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = result.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "eb1029d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>average_expression_breadth</th>\n",
       "      <th>num_species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein_S_cerevisiae</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03653</th>\n",
       "      <td>194.037491</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03697</th>\n",
       "      <td>166.221399</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03708</th>\n",
       "      <td>132.368049</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03626</th>\n",
       "      <td>124.314199</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03685</th>\n",
       "      <td>119.032086</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03641</th>\n",
       "      <td>111.726172</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03625</th>\n",
       "      <td>102.393535</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>https://omabrowser.org/oma/info/YEAST03675</th>\n",
       "      <td>71.591195</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            average_expression_breadth  \\\n",
       "Protein_S_cerevisiae                                                     \n",
       "https://omabrowser.org/oma/info/YEAST03653  194.037491                   \n",
       "https://omabrowser.org/oma/info/YEAST03697  166.221399                   \n",
       "https://omabrowser.org/oma/info/YEAST03708  132.368049                   \n",
       "https://omabrowser.org/oma/info/YEAST03626  124.314199                   \n",
       "https://omabrowser.org/oma/info/YEAST03685  119.032086                   \n",
       "https://omabrowser.org/oma/info/YEAST03641  111.726172                   \n",
       "https://omabrowser.org/oma/info/YEAST03625  102.393535                   \n",
       "https://omabrowser.org/oma/info/YEAST03675  71.591195                    \n",
       "\n",
       "                                            num_species  \n",
       "Protein_S_cerevisiae                                     \n",
       "https://omabrowser.org/oma/info/YEAST03653  40.0         \n",
       "https://omabrowser.org/oma/info/YEAST03697  26.0         \n",
       "https://omabrowser.org/oma/info/YEAST03708  28.0         \n",
       "https://omabrowser.org/oma/info/YEAST03626  33.0         \n",
       "https://omabrowser.org/oma/info/YEAST03685  28.0         \n",
       "https://omabrowser.org/oma/info/YEAST03641  29.0         \n",
       "https://omabrowser.org/oma/info/YEAST03625  27.0         \n",
       "https://omabrowser.org/oma/info/YEAST03675  14.0         "
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d9f9b4",
   "metadata": {},
   "source": [
    "**Caveat**: retrieving only partial results will naturally result in only a partial view of the datasets involved, therefore the results should not be taken at face value, but rather experiments should be repeated without the *limit*s on the queries. Currently however this is not possible due to performance issues."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
